### PR TITLE
roachtest: add support for Side-Eye

### PR DESCRIPTION
**Pull request by andreimatei**
This patch teaches roachtest to install the Side-Eye agents on the
roachprod clusters it creates, if configured. If a test times out,
roachtest asks Side-Eye to capture a snapshot of all the cockroach
processes in the cluster before terminating the test. This should prove
useful for figuring out why the test timed out / what the cluster was
doing when the timer expired.

Side-Eye (side-eye.io) is a distribute debugger that can take
"snapshots" of a Cockroach cluster, capturing all the stack traces and
variables/expressions from these stacks. It helps you "build reports"
about the state of a cluster -- for example, you can get SQL tables with
all the queries that are currently in-flight, see which one is the
oldest, see which transactions are blocked on which other transactions.
Or, you can get information on background processes -- for example you
can get info on the state of all the rangefeeds. In the not-too-far
future, we're promising all sorts of other functionality in the debugger
(dynamic logs and metrics, integration with Go's runtime tracing, CPU
profiling).

roachprod already had support for deploying the Side-Eye agents
(`roachprod install side-eye`); this patch hooks up roachtest to take
advantage of that.

The configuration is done through a new `--side-eye-token` flag to
roachtest. The value of this token can be retrieved by logging in to
[app.side-eye.io](https://app.side-eye.io) with a corp email.

When roachtests time out, they now print a link to the snapshot that was
captured. This snapshot outlive the cluster and everybody at CRL can
access it.

Also, the roachtest web UI displays links to Side-Eye for each roachprod
cluster. Each cluster corresponds to a Side-Eye "environment"; these
links navigate to the respective environment, allowing the user to
manually capture snapshots of the cluster and to see the snapshots of
the cluster that have already been captured.
The link looks like this:

Screenshot from 2024-05-28 19-38-49

and it takes you to a Side-Eye page that looks like this

Screenshot from 2024-05-28 18-19-19

Also, whenever the test runner takes a snapshot of a cluster (i.e. on
test timeout), we add an annotation about this to the cluster's Grafana
dashboards.

In addition to being used by the roachtest on timeouts (and maybe on
other conditions in the future? Any failure, perhaps?), I hope people
running benchmarks manually will take advantage of Side-Eye by finding
their clusters in the Side-Eye app, capturing snapshots at will,
refining the collected data, and learning interesting things about
cockroach's behavior.

Epic: None
Release note: None

### CODE REVIEW

**Comment by andreimatei** [comment]
@dt @msbutler @srosenberg how does this sound? I think the Side-Eye integration should prove quite useful for roachtests. I'm open to any suggestion. Thanks!

---

Only that last 3 commits are to be reviewed here. The rest are #120340

**Comment by dt** [comment]
Is the last commit all we need in this PR? or does it rely on the one before it (the timing of linking the cluster)?

**Comment by andreimatei** [comment]
> Is the last commit all we need in this PR? or does it rely on the one before it (the timing of linking the cluster)?

The linking the cluster commit is not needed. I can send it separately or even drop it if you prefer. It's only here because it made working with roachtest locally more pleasant (I was on a low Internet connection and uploading the cockroach binary to the roachprod clusters was always taking a while and I wanted the clusters to show up quicker in the UI).

**Comment by dt** [comment]
I'd suggest pulling the last commit and only the last commit into its own PR, since as-is this larger set of changes is blocked on a test-eng review and is picking up merge conflicts.

I can do that if you want since it needs an employee to run the mirroring anyway, so I can just amend it after doing so.

**Comment by dt** [comment]
https://github.com/cockroachdb/cockroach/pull/124889

**Comment by dt** [comment]
I pulled a couple of lines from this change out into a smaller initial diff over in https://github.com/cockroachdb/cockroach/pull/124889. 

That diff is far less comprehensive -- no addition to the UI, no new fields in various structs, nor the test-based env names and implied ability to update env names on a single cluster between tests.

But I think #124889 is the minimal diff we need to add some Start/Snapshot calls to one or two tests how it works out?

**Comment by andreimatei** [comment]
I've extracted the minor stuff in #124922; now there's a single commit here.
And we can let it linger for after your https://github.com/cockroachdb/cockroach/pull/124889.

**Comment by andreimatei** [comment]
Now rebased on #124889; only the last commit is original.

File: pkg/cmd/roachtest/run.go [hunk 1]
```diff
@@ -118,6 +132,32 @@ func runTests(register func(registry.Registry), filter *registry.TestFilter) err
 		opt.debugMode = NoDebug
 	}
 
+	if roachtestflags.SideEyeApiToken != "" {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/run.go:135]: For CI, we'll need to update the wrapper scripts to thread the token. E.g., see [1], which threads the value to `--use-spot`.

[1] https://github.com/cockroachdb/cockroach/pull/124258

File: pkg/cmd/roachtest/run.go [hunk 2]
```diff
@@ -118,6 +132,32 @@ func runTests(register func(registry.Registry), filter *registry.TestFilter) err
...
+		if roachtestflags.Local {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/run.go:136]: Nit: could move the body to a helper func. for readability; `runTests` is already plentiful with boilerplate.

File: pkg/cmd/roachtest/test_runner.go [hunk 3]
```diff
@@ -1273,6 +1297,11 @@ func (r *testRunner) runTest(
 		// NB: We're adding the timeout failure intentionally without cancelling the context
 		// to capture as much state as possible during artifact collection.
 		t.addFailure(0, "test timed out (%s)", timeout)
+		// If the Side-Eye integration was configured, capture a snapshot of the
+		// cluster before we destroy it.
+		if r.sideEyeClient != nil {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:1302]: Any reason this is not inside `teardownTest`? Logically, all post-processing happens there.

Comment (srosenberg) [reply pkg/cmd/roachtest/test_runner.go:1302]: In fact, `collectArtifacts` is even more fitting.

File: pkg/roachprod/roachprod.go [hunk 4]
```diff
@@ -2748,6 +2811,77 @@ func Deploy(
...
+		client, err = sideeyeclient.NewSideEyeClient(sideeyeclient.WithApiToken(sideEyeToken))
+		if err != nil {
+			l.Errorf("failed to create Side-Eye client: %s", err)
+			return "", false
+		}
+		defer client.Close()
+	}
+
+	// Protect against the snapshot taking too long.
+	snapCtx, cancel := context.WithTimeout(ctx, time.Second*30)
```
Comment (srosenberg) [inline pkg/roachprod/roachprod.go:2864]: How long does it take on average? What if it's a large cluster, would 30 seconds suffice then?

**Comment by dt** [comment]
@andreimatei https://github.com/cockroachdb/cockroach/pull/124889 is in if you wanna rebase one last time

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124780)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/run.go` line 135 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1655715399:-O0kLmke9MBLpBOz8AAd:bymqhgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/af2e05c56b40fc72a0993000e43d72f407b3d069/pkg/cmd/roachtest/run.go#L135)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

For CI, we'll need to update the wrapper scripts to thread the token. E.g., see [1], which threads the value to `--use-spot`.

[1] https://github.com/cockroachdb/cockroach/pull/124258
</blockquote></details>

I'm happy to update those scripts. But I want to talk to you about where we want this API token to come from (I'll do whatever you prefer). I  think there's an analogous `--slack-token` flag to roachtest which comes [roachtest_nightly_gce_impl.sh](https://github.com/cockroachdb/cockroach/blob/a843435384b48c3233af37744ef412921872beda/build/teamcity/cockroach/coverage/roachtest_nightly_gce_impl.sh#L36) passes from an env var. I guess the value of that env var is set directly in TC?

There's also a `--dogstatsd-addr` flag for talking to a DataDog agent (I think). I can't tell who sets that, and what secrets are involved there.

So, what should we do for our token? An env var set in TC?
Should I get rid of supporting reading the token from `gcloud secrets`?
___
*[`pkg/cmd/roachtest/run.go` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1655716159:-O0kLSR-1eOn5p81P-Op:b61k34l) ([raw file](https://github.com/cockroachdb/cockroach/blob/af2e05c56b40fc72a0993000e43d72f407b3d069/pkg/cmd/roachtest/run.go#L136)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Nit: could move the body to a helper func. for readability; `runTests` is already plentiful with boilerplate.
</blockquote></details>

done; good call
___
*[`pkg/cmd/roachtest/test_runner.go` line 1302 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1655719949:-O0kQUAR0R8C6kXSCn2z:bzdyab7) ([raw file](https://github.com/cockroachdb/cockroach/blob/af2e05c56b40fc72a0993000e43d72f407b3d069/pkg/cmd/roachtest/test_runner.go#L1302)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

In fact, `collectArtifacts` is even more fitting.
</blockquote></details>

I moved it to `teardownTest`. Didn't go as far as `collectArtifacts` because that one seems more tied to TC artifacts specifically. But I can move it if you prefer.

Now that the code is in `teardownTest`, it does beg the question of whether we want a snapshot on all failures, or just timeouts. I timidly started with timeouts, but I'll happily do it for every failure if you think so; I do think that these snapshots can show interesting things in many situations and I'd definitely want one of them whenever I get a `debug.zip`.
___
*[`pkg/roachprod/roachprod.go` line 2864 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1655720943:-O0kRS1v-U11c592CndV:b34m85d) ([raw file](https://github.com/cockroachdb/cockroach/blob/af2e05c56b40fc72a0993000e43d72f407b3d069/pkg/roachprod/roachprod.go#L2864)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

How long does it take on average? What if it's a large cluster, would 30 seconds suffice then?
</blockquote></details>

It generally takes under a second if we've seen the binary before, or around 10s I think if we haven't (so, for a random nightly, the first snapshot will take a bit longer, and then further ones are super quick). The size of the cluster doesn't matter directly -- our control plane tells all the agents to take snapshots concurrently.

But I've changed the timeout to a minute since you don't seem scared.

File: pkg/roachprod/roachprod.go [hunk 5]
```diff
@@ -2356,6 +2356,68 @@ func StopOpenTelemetry(ctx context.Context, l *logger.Logger, clusterName string
...
+	ctx context.Context, l *logger.Logger, clusterName string, envName string, apiToken string,
+) error {
+	c, err := getClusterFromCache(l, clusterName)
+	if err != nil {
+		return err
+	}
+
+	// Note that this command is similar to the one used by `roachprod install
+	// side-eye`. We could use that through install.InstallTool(), but that code
+	// is looks up the API token in `gcloud secrets`; we already know the token,
```
Comment (herkolategan) [inline pkg/roachprod/roachprod.go:2380]: Nit: small comment typo "is"

**Review by herkolategan — COMMENTED** [review]
As Stan suggested if we add passing the `SIDE_EYE_API_TOKEN` in the TC scripts. We can configure the TC job to pass it (securely), and run a TC roachtest job for this PR to confirm everything works as expected.

> So, what should we do for our token? An env var set in TC?

For our nightlies / weeklies I prefer TC passing the secret as an env var to keep it consistent.

**Review by andreimatei — COMMENTED** [review]
OK, I've updated the roachtest*.sh scripts to thread `--side-eye-token=${SIDE_EYE_API_TOKEN}`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124780#-:-O1XF9pK-tiSpXN0KSaz:b-ytcsxz)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachprod/main.go` line 1648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1673756901:-O1X0F8a3aps5sdzeeFr:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/904a5e17139fdf441c5cabc7a9c041b1b731f457/pkg/cmd/roachprod/main.go#L1648)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: traling comma in the help message.
```
❯ ./bin/roachprod side-eye install --help
Install and start the Side-Eye agents on all nodes in the cluster

`roachprod side-eye snapshot <cluster>` can then be used to capture cluster snapshots.,

Usage:
  roachprod side-eye install <cluster> [flags]
```
</blockquote></details>

done
___
*[`pkg/cmd/roachtest/run.go` line 135 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1655715399:-O1XEl0J0Hr2gkZQ2Q-G:bbrtv7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/af2e05c56b40fc72a0993000e43d72f407b3d069/pkg/cmd/roachtest/run.go#L135)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm happy to update those scripts. But I want to talk to you about where we want this API token to come from (I'll do whatever you prefer). I  think there's an analogous `--slack-token` flag to roachtest which comes [roachtest_nightly_gce_impl.sh](https://github.com/cockroachdb/cockroach/blob/a843435384b48c3233af37744ef412921872beda/build/teamcity/cockroach/coverage/roachtest_nightly_gce_impl.sh#L36) passes from an env var. I guess the value of that env var is set directly in TC?

There's also a `--dogstatsd-addr` flag for talking to a DataDog agent (I think). I can't tell who sets that, and what secrets are involved there.

So, what should we do for our token? An env var set in TC?
Should I get rid of supporting reading the token from `gcloud secrets`?
</blockquote></details>

I've updated the roachtest*.sh scripts to thread `--side-eye-token`.
___
*[`pkg/roachprod/roachprod.go` line 2380 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/124780#gh-1673835425:-O1XGTCjEJFpxEv6lC-2:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/904a5e17139fdf441c5cabc7a9c041b1b731f457/pkg/roachprod/roachprod.go#L2380)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Nit: small comment typo "is"
</blockquote></details>

done

**Review by andreimatei — COMMENTED** [review]
I removed the code around `roachtest` reading the api token from `gcloud secret`, since it doesn't look like we'll use that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124780#-:-O1XGiB-1DDAxpOXj2hR:blp82g4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

**Review by srosenberg — APPROVED** [review]
Looks good! Let me kick off a smoke test. Barring any surprises, I think we can merge tomorrow.

**Comment by srosenberg** [comment]
[SELECT_PROBABILITY=0.4](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/16040411?buildTab=log&logFilter=debug&focusLine=6&logView=linear)

Note I used the same token we have under `gcloud secrets`. For post-merge runs, did you want to use a different one?

**Comment by srosenberg** [comment]
Mostly VM preemptions in the smoke test, as expected. I've updated [Roachtest Nightly GCE](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel?buildTypeTab=buildTypeSettings#all-projects) with a new token (as instructed by @andreimatei via DM).

bors r+

**Pull request merged**